diff --git a/src/common/runtime.cpp b/src/common/runtime.cpp
--- a/src/common/runtime.cpp
+++ b/src/common/runtime.cpp
@@ -647,5 +647,12 @@ void luax_rawnewtype(lua_State *L, love::Type &type, love::Object *object)
 {
+#ifndef LOVE_PROXY_USERVALUES
+#define LOVE_PROXY_USERVALUES 1
+#endif
+#if LOVE_PROXY_USERVALUES > 1
+	Proxy *u = (Proxy *)lua_newuserdatauv(L, sizeof(Proxy), LOVE_PROXY_USERVALUES);
+#else
 	Proxy *u = (Proxy *)lua_newuserdata(L, sizeof(Proxy));
+#endif
 
 	object->retain();
 
