--- a/src/modules/graphics/wrap_Graphics.cpp
+++ b/src/modules/graphics/wrap_Graphics.cpp
@@ -551,63 +551,10 @@
 		[&]() { instance()->captureScreenshot(info); },
 		[&](bool except) { if (except) info.callback(&info, nullptr, nullptr); }
 	);
-
-	return 0;
-}
-
-int w_setScissor(lua_State *L)
-{
-	int nargs = lua_gettop(L);
-
-	if (nargs == 0 || (nargs == 4 && lua_isnil(L, 1) && lua_isnil(L, 2)
-		&& lua_isnil(L, 3) && lua_isnil(L, 4)))
-	{
-		instance()->setScissor();
-		return 0;
-	}
 
-	Rect rect;
-	rect.x = (int) luaL_checkinteger(L, 1);
-	rect.y = (int) luaL_checkinteger(L, 2);
-	rect.w = (int) luaL_checkinteger(L, 3);
-	rect.h = (int) luaL_checkinteger(L, 4);
-
-	if (rect.w < 0 || rect.h < 0)
-		return luaL_error(L, "Can't set scissor with negative width and/or height.");
-
-	instance()->setScissor(rect);
 	return 0;
 }
 
-int w_intersectScissor(lua_State *L)
-{
-	Rect rect;
-	rect.x = (int) luaL_checkinteger(L, 1);
-	rect.y = (int) luaL_checkinteger(L, 2);
-	rect.w = (int) luaL_checkinteger(L, 3);
-	rect.h = (int) luaL_checkinteger(L, 4);
-
-	if (rect.w < 0 || rect.h < 0)
-		return luaL_error(L, "Can't set scissor with negative width and/or height.");
-
-	instance()->intersectScissor(rect);
-	return 0;
-}
-
-int w_getScissor(lua_State *L)
-{
-	Rect rect;
-	if (!instance()->getScissor(rect))
-		return 0;
-
-	lua_pushinteger(L, rect.x);
-	lua_pushinteger(L, rect.y);
-	lua_pushinteger(L, rect.w);
-	lua_pushinteger(L, rect.h);
-
-	return 4;
-}
-
 int w_stencil(lua_State *L)
 {
 	luaL_checktype(L, 1, LUA_TFUNCTION);
--- a/src/modules/graphics/wrap_GraphicsDisplayState.h
+++ b/src/modules/graphics/wrap_GraphicsDisplayState.h
@@ -21,6 +21,7 @@
 #pragma once
 
 #include "common/Color.h"
+#include "common/math.h"
 #include "common/runtime.h"
 
 namespace love
@@ -66,6 +67,10 @@
 	virtual ColorMask getColorMask() const = 0;
 	virtual void setWireframe(bool enabled) = 0;
 	virtual bool isWireframe() const = 0;
+	virtual void setScissor() = 0;
+	virtual void setScissor(Rect rect) = 0;
+	virtual void intersectScissor(Rect rect) = 0;
+	virtual bool getScissor(Rect &rect) const = 0;
 };
 
 int w_setColor(lua_State *L);
@@ -84,6 +89,9 @@
 int w_getColorMask(lua_State *L);
 int w_setWireframe(lua_State *L);
 int w_isWireframe(lua_State *L);
+int w_setScissor(lua_State *L);
+int w_intersectScissor(lua_State *L);
+int w_getScissor(lua_State *L);
 
 } // graphics
 } // love
--- a/src/modules/graphics/wrap_GraphicsDisplayState.cpp
+++ b/src/modules/graphics/wrap_GraphicsDisplayState.cpp
@@ -211,5 +211,50 @@
 	return 1;
 }
 
+int w_setScissor(lua_State *L)
+{
+	const int nargs = lua_gettop(L);
+	if (nargs == 0 || (nargs == 4 && lua_isnil(L, 1) && lua_isnil(L, 2)
+		&& lua_isnil(L, 3) && lua_isnil(L, 4)))
+	{
+		displayStateCommand(L)->setScissor();
+		return 0;
+	}
+	Rect rect;
+	rect.x = (int) luaL_checkinteger(L, 1);
+	rect.y = (int) luaL_checkinteger(L, 2);
+	rect.w = (int) luaL_checkinteger(L, 3);
+	rect.h = (int) luaL_checkinteger(L, 4);
+	if (rect.w < 0 || rect.h < 0)
+		return luaL_error(L, "Can't set scissor with negative width and/or height.");
+	displayStateCommand(L)->setScissor(rect);
+	return 0;
+}
+
+int w_intersectScissor(lua_State *L)
+{
+	Rect rect;
+	rect.x = (int) luaL_checkinteger(L, 1);
+	rect.y = (int) luaL_checkinteger(L, 2);
+	rect.w = (int) luaL_checkinteger(L, 3);
+	rect.h = (int) luaL_checkinteger(L, 4);
+	if (rect.w < 0 || rect.h < 0)
+		return luaL_error(L, "Can't set scissor with negative width and/or height.");
+	displayStateCommand(L)->intersectScissor(rect);
+	return 0;
+}
+
+int w_getScissor(lua_State *L)
+{
+	Rect rect;
+	if (!displayStateCommand(L)->getScissor(rect))
+		return 0;
+	lua_pushinteger(L, rect.x);
+	lua_pushinteger(L, rect.y);
+	lua_pushinteger(L, rect.w);
+	lua_pushinteger(L, rect.h);
+	return 4;
+}
+
 } // graphics
 } // love

