--- a/src/modules/graphics/wrap_Graphics.cpp
+++ b/src/modules/graphics/wrap_Graphics.cpp
@@ -592,42 +592,7 @@
 	luax_catchexcept(L, [&](){ instance()->stopDrawToStencilBuffer(); });
 	return 0;
 }
-
-int w_setStencilTest(lua_State *L)
-{
-	// COMPARE_ALWAYS effectively disables stencil testing.
-	CompareMode compare = COMPARE_ALWAYS;
-	int comparevalue = 0;
 
-	if (!lua_isnoneornil(L, 1))
-	{
-		const char *comparestr = luaL_checkstring(L, 1);
-		if (!getConstant(comparestr, compare))
-			return luax_enumerror(L, "compare mode", getConstants(compare), comparestr);
-
-		comparevalue = (int) luaL_checkinteger(L, 2);
-	}
-
-	luax_catchexcept(L, [&](){ instance()->setStencilTest(compare, comparevalue); });
-	return 0;
-}
-
-int w_getStencilTest(lua_State *L)
-{
-	CompareMode compare = COMPARE_ALWAYS;
-	int comparevalue = 1;
-
-	instance()->getStencilTest(compare, comparevalue);
-
-	const char *comparestr;
-	if (!getConstant(compare, comparestr))
-		return luaL_error(L, "Unknown compare mode.");
-
-	lua_pushstring(L, comparestr);
-	lua_pushnumber(L, comparevalue);
-	return 2;
-}
-
 static void parseDPIScale(Data *d, float *dpiscale)
 {
 	auto fd = dynamic_cast<love::filesystem::FileData *>(d);
--- a/src/modules/graphics/wrap_GraphicsDisplayState.h
+++ b/src/modules/graphics/wrap_GraphicsDisplayState.h
@@ -129,6 +129,8 @@
 	virtual CullMode getMeshCullMode() const = 0;
 	virtual void setFrontFaceWinding(Winding winding) = 0;
 	virtual Winding getFrontFaceWinding() const = 0;
+	virtual void setStencilTest(CompareMode compare, int value) = 0;
+	virtual CompareMode getStencilTest(int &value) const = 0;
 };
 
 int w_setColor(lua_State *L);
@@ -162,6 +164,8 @@
 int w_getMeshCullMode(lua_State *L);
 int w_setFrontFaceWinding(lua_State *L);
 int w_getFrontFaceWinding(lua_State *L);
+int w_setStencilTest(lua_State *L);
+int w_getStencilTest(lua_State *L);
 
 } // graphics
 } // love
--- a/src/modules/graphics/wrap_GraphicsDisplayState.cpp
+++ b/src/modules/graphics/wrap_GraphicsDisplayState.cpp
@@ -500,5 +500,32 @@
 	return 1;
 }
 
+int w_setStencilTest(lua_State *L)
+{
+	auto compare = GraphicsDisplayStateCommand::COMPARE_ALWAYS;
+	int value = 0;
+	if (!lua_isnoneornil(L, 1))
+	{
+		const char *str = luaL_checkstring(L, 1);
+		if (!getCompareConstant(str, compare))
+			return luax_enumerror(L, "compare mode", std::vector<std::string>{"less", "lequal",
+				"equal", "gequal", "greater", "notequal", "always", "never"}, str);
+		value = (int) luaL_checkinteger(L, 2);
+	}
+	displayStateCommand(L)->setStencilTest(compare, value);
+	return 0;
+}
+
+int w_getStencilTest(lua_State *L)
+{
+	int value;
+	const char *str = getCompareConstant(displayStateCommand(L)->getStencilTest(value));
+	if (str == nullptr)
+		return luaL_error(L, "Unknown compare mode.");
+	lua_pushstring(L, str);
+	lua_pushnumber(L, value);
+	return 2;
+}
+
 } // graphics
 } // love

