--- a/src/modules/graphics/wrap_Graphics.cpp
+++ b/src/modules/graphics/wrap_Graphics.cpp
@@ -64,18 +64,6 @@
 {
 	if (!instance()->isCreated())
 		return luaL_error(L, "love.graphics cannot function without a window!");
-	return 0;
-}
-
-int w_reset(lua_State *)
-{
-	instance()->reset();
-	return 0;
-}
-
-int w_present(lua_State *L)
-{
-	luax_catchexcept(L, [&]() { instance()->present(L); });
 	return 0;
 }
 
@@ -1934,12 +1922,6 @@
 	coords[numvertices] = coords[0];
 
 	luax_catchexcept(L, [&](){ instance()->polygon(mode, coords, numvertices+1); });
-	return 0;
-}
-
-int w_flushBatch(lua_State *)
-{
-	instance()->flushStreamDraws();
 	return 0;
 }
 
--- a/src/modules/graphics/wrap_GraphicsState.h
+++ b/src/modules/graphics/wrap_GraphicsState.h
@@ -41,6 +41,9 @@
 		STACK_ALL,
 	};
 	virtual ~GraphicsStateCommand() = default;
+	virtual void reset(lua_State *L) = 0;
+	virtual void present(lua_State *L) = 0;
+	virtual void flushBatch() = 0;
 	virtual size_t getStackDepth() const = 0;
 	virtual void push(lua_State *L, StackType type) = 0;
 	virtual void pop(lua_State *L) = 0;
@@ -55,6 +58,9 @@
 	virtual Vector2 inverseTransformPoint(Vector2 point) const = 0;
 };
 
+int w_reset(lua_State *L);
+int w_present(lua_State *L);
+int w_flushBatch(lua_State *L);
 int w_getStackDepth(lua_State *L);
 int w_push(lua_State *L);
 int w_pop(lua_State *L);
--- a/src/modules/graphics/wrap_GraphicsState.cpp
+++ b/src/modules/graphics/wrap_GraphicsState.cpp
@@ -42,6 +42,24 @@
 	return command;
 }
 
+int w_reset(lua_State *L)
+{
+	stateCommand(L)->reset(L);
+	return 0;
+}
+
+int w_present(lua_State *L)
+{
+	luax_catchexcept(L, [&]() { stateCommand(L)->present(L); });
+	return 0;
+}
+
+int w_flushBatch(lua_State *L)
+{
+	stateCommand(L)->flushBatch();
+	return 0;
+}
+
 int w_getStackDepth(lua_State *L)
 {
 	lua_pushnumber(L, stateCommand(L)->getStackDepth());

