--- a/src/modules/graphics/wrap_Mesh.cpp
+++ b/src/modules/graphics/wrap_Mesh.cpp
@@ -418,7 +418,7 @@
 		Data *d = luax_totype<Data>(L, 2, Data::type);
 
 		const char *indextypestr = luaL_checkstring(L, 3);
-		IndexDataType indextype;
+		IndexDataType indextype = INDEX_UINT16;
 		if (!vertex::getConstant(indextypestr, indextype))
 			return luax_enumerror(L, "index data type", vertex::getConstants(indextype), indextypestr);
 
@@ -523,7 +523,7 @@
 {
 	Mesh *t = luax_checkmesh(L, 1);
 	const char *str = luaL_checkstring(L, 2);
-	PrimitiveType mode;
+	PrimitiveType mode = PRIMITIVE_TRIANGLES;
 
 	if (!vertex::getConstant(str, mode))
 		return luax_enumerror(L, "mesh draw mode", vertex::getConstants(mode), str);

