diff --git a/src/modules/graphics/wrap_GraphicsCapabilities.h b/src/modules/graphics/wrap_GraphicsCapabilities.h
index bb5660c06..4bba0067a 100644
--- a/src/modules/graphics/wrap_GraphicsCapabilities.h
+++ b/src/modules/graphics/wrap_GraphicsCapabilities.h
@@ -23,6 +23,7 @@ class GraphicsCapabilitiesCommand
 	virtual BoolFields getSupported() const = 0;
 	virtual BoolFields getTextureTypes() const = 0;
 	virtual BoolFields getImageFormats() const = 0;
+	virtual BoolFields getCanvasFormats(bool readable) const = 0;
 	virtual RendererInfo getRendererInfo() const = 0;
 	virtual NumberFields getSystemLimits() const = 0;
 	virtual Stats getStats() const = 0;
@@ -31,6 +32,7 @@ class GraphicsCapabilitiesCommand
 LOVE_EXPORT int w_getSupported(lua_State *L);
 LOVE_EXPORT int w_getTextureTypes(lua_State *L);
 LOVE_EXPORT int w_getImageFormats(lua_State *L);
+LOVE_EXPORT int w_getCanvasFormats(lua_State *L);
 LOVE_EXPORT int w_getRendererInfo(lua_State *L);
 LOVE_EXPORT int w_getSystemLimits(lua_State *L);
 LOVE_EXPORT int w_getStats(lua_State *L);
diff --git a/src/modules/graphics/wrap_GraphicsCapabilities.cpp b/src/modules/graphics/wrap_GraphicsCapabilities.cpp
index c4aa36d80..f428b639d 100644
--- a/src/modules/graphics/wrap_GraphicsCapabilities.cpp
+++ b/src/modules/graphics/wrap_GraphicsCapabilities.cpp
@@ -50,6 +50,27 @@ int w_getImageFormats(lua_State *L)
 	return pushBoolFields(L, fields);
 }
 
+int w_getCanvasFormats(lua_State *L)
+{
+	bool readable = true;
+	int tableIndex = 1;
+	if (lua_type(L, 1) == LUA_TBOOLEAN)
+	{
+		readable = luax_checkboolean(L, 1);
+		tableIndex = 2;
+	}
+	GraphicsCapabilitiesCommand::BoolFields fields;
+	luax_catchexcept(L, [&](){ fields = capabilitiesCommand(L)->getCanvasFormats(readable); });
+	if (lua_istable(L, tableIndex)) lua_pushvalue(L, tableIndex);
+	else lua_createtable(L, 0, static_cast<int>(fields.size()));
+	for (const auto &[name, value] : fields)
+	{
+		luax_pushboolean(L, value);
+		lua_setfield(L, -2, name.c_str());
+	}
+	return 1;
+}
+
 int w_getRendererInfo(lua_State *L)
 {
 	GraphicsCapabilitiesCommand::RendererInfo info;
