diff --git a/Source/3rdParty/Love/src/modules/graphics/wrap_GraphicsImageConstructor.h b/Source/3rdParty/Love/src/modules/graphics/wrap_GraphicsImageConstructor.h
new file mode 100644
index 000000000..78969e7d6
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/graphics/wrap_GraphicsImageConstructor.h
@@ -0,0 +1,22 @@
+/** Copyright (c) 2006-2023 LOVE Development Team. */
+#pragma once
+
+#include "Image.h"
+#include "common/runtime.h"
+
+namespace love::graphics
+{
+
+class GraphicsImageConstructorCommand
+{
+public:
+	virtual ~GraphicsImageConstructorCommand() = default;
+	virtual Image *newImage(Image::Slices &slices, const Image::Settings &settings) = 0;
+};
+
+LOVE_EXPORT int w_newImage(lua_State *L);
+LOVE_EXPORT int w_newArrayImage(lua_State *L);
+LOVE_EXPORT int w_newCubeImage(lua_State *L);
+LOVE_EXPORT int w_newVolumeImage(lua_State *L);
+
+}
diff --git a/Source/3rdParty/Love/src/modules/graphics/wrap_GraphicsImageConstructor.cpp b/Source/3rdParty/Love/src/modules/graphics/wrap_GraphicsImageConstructor.cpp
new file mode 100644
index 000000000..201d670ae
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/graphics/wrap_GraphicsImageConstructor.cpp
@@ -0,0 +1,383 @@
+/** Copyright (c) 2006-2023 LOVE Development Team. */
+#include "wrap_GraphicsImageConstructor.h"
+
+#include "common/Module.h"
+#include "filesystem/Filesystem.h"
+#include "filesystem/wrap_Filesystem.h"
+#include "image/Image.h"
+#include "image/wrap_Image.h"
+
+#include <algorithm>
+#include <cstdlib>
+
+namespace love::graphics
+{
+
+static GraphicsImageConstructorCommand *imageConstructorCommand(lua_State *L)
+{
+	auto *module = luax_getmodule(L, Module::M_GRAPHICS);
+	auto *command = dynamic_cast<GraphicsImageConstructorCommand *>(module);
+	if (!command) luaL_error(L, "love.graphics has no state-local Image factory");
+	return command;
+}
+
+static love::image::Image *imageModule(lua_State *L)
+{
+	auto *module = luax_getmodule(L, Module::M_IMAGE);
+	auto *image = dynamic_cast<love::image::Image *>(module);
+	if (!image) luaL_error(L, "Cannot load images without the love.image module.");
+	return image;
+}
+
+static void parseDPIScale(Data *d, float *dpiscale)
+{
+	auto fd = dynamic_cast<love::filesystem::FileData *>(d);
+	if (fd == nullptr)
+		return;
+
+	// Parse a density scale of 2.0 from "image@2x.png".
+	const std::string &fname = fd->getName();
+
+	size_t namelen = fname.length();
+	size_t atpos = fname.rfind('@');
+
+	if (atpos != std::string::npos && atpos + 2 < namelen
+		&& (fname[namelen - 1] == 'x' || fname[namelen - 1] == 'X'))
+	{
+		char *end = nullptr;
+		long density = strtol(fname.c_str() + atpos + 1, &end, 10);
+		if (end != nullptr && density > 0 && dpiscale != nullptr)
+			*dpiscale = (float) density;
+	}
+}
+
+static Image::Settings w__optImageSettings(lua_State *L, int idx, bool &setdpiscale)
+{
+	Image::Settings s;
+
+	setdpiscale = false;
+	if (!lua_isnoneornil(L, idx))
+	{
+		luax_checktablefields<Image::SettingType>(L, idx, "image setting name", Image::getConstant);
+
+		s.mipmaps = luax_boolflag(L, idx, Image::getConstant(Image::SETTING_MIPMAPS), s.mipmaps);
+		s.linear = luax_boolflag(L, idx, Image::getConstant(Image::SETTING_LINEAR), s.linear);
+
+		lua_getfield(L, idx, Image::getConstant(Image::SETTING_DPI_SCALE));
+		if (lua_isnumber(L, -1))
+		{
+			s.dpiScale = (float) lua_tonumber(L, -1);
+			setdpiscale = true;
+		}
+		lua_pop(L, 1);
+	}
+
+	return s;
+}
+
+static std::pair<StrongRef<image::ImageData>, StrongRef<image::CompressedImageData>>
+getImageData(lua_State *L, int idx, bool allowcompressed, float *dpiscale)
+{
+	StrongRef<image::ImageData> idata;
+	StrongRef<image::CompressedImageData> cdata;
+
+	if (luax_istype(L, idx, image::ImageData::type))
+		idata.set(image::luax_checkimagedata(L, idx));
+	else if (luax_istype(L, idx, image::CompressedImageData::type))
+		cdata.set(image::luax_checkcompressedimagedata(L, idx));
+	else if (filesystem::luax_cangetdata(L, idx))
+	{
+		// Convert to ImageData / CompressedImageData.
+		auto imagemodule = imageModule(L);
+		if (imagemodule == nullptr)
+			luaL_error(L, "Cannot load images without the love.image module.");
+
+		StrongRef<Data> fdata(filesystem::luax_getdata(L, idx), Acquire::NORETAIN);
+
+		if (dpiscale != nullptr)
+			parseDPIScale(fdata, dpiscale);
+
+		if (allowcompressed && imagemodule->isCompressed(fdata))
+			luax_catchexcept(L, [&]() { cdata.set(imagemodule->newCompressedData(fdata), Acquire::NORETAIN); });
+		else
+			luax_catchexcept(L, [&]() { idata.set(imagemodule->newImageData(fdata), Acquire::NORETAIN); });
+	}
+	else
+		idata.set(image::luax_checkimagedata(L, idx));
+
+	return std::make_pair(idata, cdata);
+}
+
+static int w__pushNewImage(lua_State *L, Image::Slices &slices, const Image::Settings &settings)
+{
+	StrongRef<Image> i;
+	luax_catchexcept(L,
+		[&]() { i.set(imageConstructorCommand(L)->newImage(slices, settings), Acquire::NORETAIN); },
+		[&](bool) { slices.clear(); }
+	);
+
+	luax_pushtype(L, i);
+	return 1;
+}
+
+int w_newCubeImage(lua_State *L)
+{
+	Image::Slices slices(TEXTURE_CUBE);
+
+	bool dpiscaleset = false;
+	Image::Settings settings = w__optImageSettings(L, 2, dpiscaleset);
+	float *autodpiscale = dpiscaleset ? nullptr : &settings.dpiScale;
+
+	auto imagemodule = imageModule(L);
+
+	if (!lua_istable(L, 1))
+	{
+		auto data = getImageData(L, 1, true, autodpiscale);
+
+		std::vector<StrongRef<love::image::ImageData>> faces;
+
+		if (data.first.get())
+		{
+			luax_catchexcept(L, [&](){ faces = imagemodule->newCubeFaces(data.first); });
+
+			for (int i = 0; i < (int) faces.size(); i++)
+				slices.set(i, 0, faces[i]);
+		}
+		else
+			slices.add(data.second, 0, 0, true, settings.mipmaps);
+	}
+	else
+	{
+		int tlen = (int) luax_objlen(L, 1);
+
+		if (luax_isarrayoftables(L, 1))
+		{
+			if (tlen != 6)
+				return luaL_error(L, "Cubemap images must have 6 faces.");
+
+			for (int face = 0; face < tlen; face++)
+			{
+				lua_rawgeti(L, 1, face + 1);
+				luaL_checktype(L, -1, LUA_TTABLE);
+
+				int miplen = std::max(1, (int) luax_objlen(L, -1));
+
+				for (int mip = 0; mip < miplen; mip++)
+				{
+					lua_rawgeti(L, -1, mip + 1);
+
+					auto data = getImageData(L, -1, true, face == 0 && mip == 0 ? autodpiscale : nullptr);
+					if (data.first.get())
+						slices.set(face, mip, data.first);
+					else
+						slices.set(face, mip, data.second->getSlice(0, 0));
+
+					lua_pop(L, 1);
+				}
+			}
+		}
+		else
+		{
+			bool usemipmaps = false;
+
+			for (int i = 0; i < tlen; i++)
+			{
+				lua_rawgeti(L, 1, i + 1);
+
+				auto data = getImageData(L, -1, true, i == 0 ? autodpiscale : nullptr);
+
+				if (data.first.get())
+				{
+					if (usemipmaps || data.first->getWidth() != data.first->getHeight())
+					{
+						usemipmaps = true;
+
+						std::vector<StrongRef<love::image::ImageData>> faces;
+						luax_catchexcept(L, [&](){ faces = imagemodule->newCubeFaces(data.first); });
+
+						for (int face = 0; face < (int) faces.size(); face++)
+							slices.set(face, i, faces[i]);
+					}
+					else
+						slices.set(i, 0, data.first);
+				}
+				else
+					slices.add(data.second, i, 0, false, settings.mipmaps);
+			}
+		}
+
+		lua_pop(L, tlen);
+	}
+
+	return w__pushNewImage(L, slices, settings);
+}
+
+int w_newArrayImage(lua_State *L)
+{
+	Image::Slices slices(TEXTURE_2D_ARRAY);
+
+	bool dpiscaleset = false;
+	Image::Settings settings = w__optImageSettings(L, 2, dpiscaleset);
+	float *autodpiscale = dpiscaleset ? nullptr : &settings.dpiScale;
+
+	if (lua_istable(L, 1))
+	{
+		int tlen = std::max(1, (int) luax_objlen(L, 1));
+
+		if (luax_isarrayoftables(L, 1))
+		{
+			for (int slice = 0; slice < tlen; slice++)
+			{
+				lua_rawgeti(L, 1, slice + 1);
+				luaL_checktype(L, -1, LUA_TTABLE);
+
+				int miplen = std::max(1, (int) luax_objlen(L, -1));
+
+				for (int mip = 0; mip < miplen; mip++)
+				{
+					lua_rawgeti(L, -1, mip + 1);
+
+					auto data = getImageData(L, -1, true, slice == 0 && mip == 0 ? autodpiscale : nullptr);
+					if (data.first.get())
+						slices.set(slice, mip, data.first);
+					else
+						slices.set(slice, mip, data.second->getSlice(0, 0));
+
+					lua_pop(L, 1);
+				}
+			}
+		}
+		else
+		{
+			for (int slice = 0; slice < tlen; slice++)
+			{
+				lua_rawgeti(L, 1, slice + 1);
+				auto data = getImageData(L, -1, true, slice == 0 ? autodpiscale : nullptr);
+				if (data.first.get())
+					slices.set(slice, 0, data.first);
+				else
+					slices.add(data.second, slice, 0, false, settings.mipmaps);
+			}
+		}
+
+		lua_pop(L, tlen);
+	}
+	else
+	{
+		auto data = getImageData(L, 1, true, autodpiscale);
+		if (data.first.get())
+			slices.set(0, 0, data.first);
+		else
+			slices.add(data.second, 0, 0, true, settings.mipmaps);
+	}
+
+	return w__pushNewImage(L, slices, settings);
+}
+
+int w_newVolumeImage(lua_State *L)
+{
+	auto imagemodule = imageModule(L);
+
+	Image::Slices slices(TEXTURE_VOLUME);
+
+	bool dpiscaleset = false;
+	Image::Settings settings = w__optImageSettings(L, 2, dpiscaleset);
+	float *autodpiscale = dpiscaleset ? nullptr : &settings.dpiScale;
+
+	if (lua_istable(L, 1))
+	{
+		int tlen = std::max(1, (int) luax_objlen(L, 1));
+
+		if (luax_isarrayoftables(L, 1))
+		{
+			for (int mip = 0; mip < tlen; mip++)
+			{
+				lua_rawgeti(L, 1, mip + 1);
+				luaL_checktype(L, -1, LUA_TTABLE);
+
+				int slicelen = std::max(1, (int) luax_objlen(L, -1));
+
+				for (int slice = 0; slice < slicelen; slice++)
+				{
+					lua_rawgeti(L, -1, slice + 1);
+
+					auto data = getImageData(L, -1, true, slice == 0 && mip == 0 ? autodpiscale : nullptr);
+					if (data.first.get())
+						slices.set(slice, mip, data.first);
+					else
+						slices.set(slice, mip, data.second->getSlice(0, 0));
+
+					lua_pop(L, 1);
+				}
+			}
+		}
+		else
+		{
+			for (int layer = 0; layer < tlen; layer++)
+			{
+				lua_rawgeti(L, 1, layer + 1);
+				auto data = getImageData(L, -1, true, layer == 0 ? autodpiscale : nullptr);
+				if (data.first.get())
+					slices.set(layer, 0, data.first);
+				else
+					slices.add(data.second, layer, 0, false, settings.mipmaps);
+			}
+		}
+
+		lua_pop(L, tlen);
+	}
+	else
+	{
+		auto data = getImageData(L, 1, true, autodpiscale);
+
+		if (data.first.get())
+		{
+			std::vector<StrongRef<love::image::ImageData>> layers;
+			luax_catchexcept(L, [&](){ layers = imagemodule->newVolumeLayers(data.first); });
+
+			for (int i = 0; i < (int) layers.size(); i++)
+				slices.set(i, 0, layers[i]);
+		}
+		else
+			slices.add(data.second, 0, 0, true, settings.mipmaps);
+	}
+
+	return w__pushNewImage(L, slices, settings);
+}
+
+int w_newImage(lua_State *L)
+{
+	Image::Slices slices(TEXTURE_2D);
+
+	bool dpiscaleset = false;
+	Image::Settings settings = w__optImageSettings(L, 2, dpiscaleset);
+	float *autodpiscale = dpiscaleset ? nullptr : &settings.dpiScale;
+
+	if (lua_istable(L, 1))
+	{
+		int n = std::max(1, (int) luax_objlen(L, 1));
+		for (int i = 0; i < n; i++)
+		{
+			lua_rawgeti(L, 1, i + 1);
+			auto data = getImageData(L, -1, true, i == 0 ? autodpiscale : nullptr);
+			if (data.first.get())
+				slices.set(0, i, data.first);
+			else
+				slices.set(0, i, data.second->getSlice(0, 0));
+		}
+		lua_pop(L, n);
+	}
+	else
+	{
+		auto data = getImageData(L, 1, true, autodpiscale);
+		if (data.first.get())
+			slices.set(0, 0, data.first);
+		else
+			slices.add(data.second, 0, 0, false, settings.mipmaps);
+	}
+
+	return w__pushNewImage(L, slices, settings);
+}
+
+
+}
