--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/joystick/Joystick.cpp
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+// LOVE
+#include "Joystick.h"
+
+// STL
+#include <cmath>
+
+namespace love
+{
+namespace joystick
+{
+
+love::Type Joystick::type("Joystick", &Object::type);
+
+float Joystick::clampval(float x)
+{
+	if (fabsf(x) < 0.01)
+		return 0.0f;
+
+	if (x < -0.99f) return -1.0f;
+	if (x > 0.99f) return 1.0f;
+
+	return x;
+}
+
+bool Joystick::getConstant(const char *in, Joystick::Hat &out)
+{
+	return hats.find(in, out);
+}
+
+bool Joystick::getConstant(Joystick::Hat in, const char *&out)
+{
+	return hats.find(in, out);
+}
+
+bool Joystick::getConstant(const char *in, Joystick::GamepadAxis &out)
+{
+	return gpAxes.find(in, out);
+}
+
+bool Joystick::getConstant(Joystick::GamepadAxis in, const char *&out)
+{
+	return gpAxes.find(in, out);
+}
+
+bool Joystick::getConstant(const char *in, Joystick::GamepadButton &out)
+{
+	return gpButtons.find(in, out);
+}
+
+bool Joystick::getConstant(Joystick::GamepadButton in, const char *&out)
+{
+	return gpButtons.find(in, out);
+}
+
+bool Joystick::getConstant(const char *in, Joystick::InputType &out)
+{
+	return inputTypes.find(in, out);
+}
+
+bool Joystick::getConstant(Joystick::InputType in, const char *&out)
+{
+	return inputTypes.find(in, out);
+}
+
+StringMap<Joystick::Hat, Joystick::HAT_MAX_ENUM>::Entry Joystick::hatEntries[] =
+{
+	{"c", Joystick::HAT_CENTERED},
+	{"u", Joystick::HAT_UP},
+	{"r", Joystick::HAT_RIGHT},
+	{"d", Joystick::HAT_DOWN},
+	{"l", Joystick::HAT_LEFT},
+	{"ru", Joystick::HAT_RIGHTUP},
+	{"rd", Joystick::HAT_RIGHTDOWN},
+	{"lu", Joystick::HAT_LEFTUP},
+	{"ld", Joystick::HAT_LEFTDOWN},
+};
+
+StringMap<Joystick::Hat, Joystick::HAT_MAX_ENUM> Joystick::hats(Joystick::hatEntries, sizeof(Joystick::hatEntries));
+
+StringMap<Joystick::GamepadAxis, Joystick::GAMEPAD_AXIS_MAX_ENUM>::Entry Joystick::gpAxisEntries[] =
+{
+	{"leftx", GAMEPAD_AXIS_LEFTX},
+	{"lefty", GAMEPAD_AXIS_LEFTY},
+	{"rightx", GAMEPAD_AXIS_RIGHTX},
+	{"righty", GAMEPAD_AXIS_RIGHTY},
+	{"triggerleft", GAMEPAD_AXIS_TRIGGERLEFT},
+	{"triggerright", GAMEPAD_AXIS_TRIGGERRIGHT},
+};
+
+StringMap<Joystick::GamepadAxis, Joystick::GAMEPAD_AXIS_MAX_ENUM> Joystick::gpAxes(Joystick::gpAxisEntries, sizeof(Joystick::gpAxisEntries));
+
+StringMap<Joystick::GamepadButton, Joystick::GAMEPAD_BUTTON_MAX_ENUM>::Entry Joystick::gpButtonEntries[] =
+{
+	{"a", GAMEPAD_BUTTON_A},
+	{"b", GAMEPAD_BUTTON_B},
+	{"x", GAMEPAD_BUTTON_X},
+	{"y", GAMEPAD_BUTTON_Y},
+	{"back", GAMEPAD_BUTTON_BACK},
+	{"guide", GAMEPAD_BUTTON_GUIDE},
+	{"start", GAMEPAD_BUTTON_START},
+	{"leftstick", GAMEPAD_BUTTON_LEFTSTICK},
+	{"rightstick", GAMEPAD_BUTTON_RIGHTSTICK},
+	{"leftshoulder", GAMEPAD_BUTTON_LEFTSHOULDER},
+	{"rightshoulder", GAMEPAD_BUTTON_RIGHTSHOULDER},
+	{"dpup", GAMEPAD_BUTTON_DPAD_UP},
+	{"dpdown", GAMEPAD_BUTTON_DPAD_DOWN},
+	{"dpleft", GAMEPAD_BUTTON_DPAD_LEFT},
+	{"dpright", GAMEPAD_BUTTON_DPAD_RIGHT},
+};
+
+StringMap<Joystick::GamepadButton, Joystick::GAMEPAD_BUTTON_MAX_ENUM> Joystick::gpButtons(Joystick::gpButtonEntries, sizeof(Joystick::gpButtonEntries));
+
+StringMap<Joystick::InputType, Joystick::INPUT_TYPE_MAX_ENUM>::Entry Joystick::inputTypeEntries[] =
+{
+	{"axis", Joystick::INPUT_TYPE_AXIS},
+	{"button", Joystick::INPUT_TYPE_BUTTON},
+	{"hat", Joystick::INPUT_TYPE_HAT},
+};
+
+StringMap<Joystick::InputType, Joystick::INPUT_TYPE_MAX_ENUM> Joystick::inputTypes(Joystick::inputTypeEntries, sizeof(Joystick::inputTypeEntries));
+
+} // joystick
+} // love
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/joystick/Joystick.h
@@ -0,0 +1,205 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#ifndef LOVE_JOYSTICK_JOYSTICK_H
+#define LOVE_JOYSTICK_JOYSTICK_H
+
+// LOVE
+#include "common/Object.h"
+#include "common/StringMap.h"
+
+// stdlib
+#include <vector>
+#include <string>
+
+namespace love
+{
+namespace joystick
+{
+
+class Joystick : public Object
+{
+public:
+
+	static love::Type type;
+
+	// Joystick hat values.
+	enum Hat
+	{
+		HAT_INVALID,
+		HAT_CENTERED,
+		HAT_UP,
+		HAT_RIGHT,
+		HAT_DOWN,
+		HAT_LEFT,
+		HAT_RIGHTUP,
+		HAT_RIGHTDOWN,
+		HAT_LEFTUP,
+		HAT_LEFTDOWN,
+		HAT_MAX_ENUM = 16
+	};
+
+	// Valid Gamepad axes.
+	enum GamepadAxis
+	{
+		GAMEPAD_AXIS_INVALID,
+		GAMEPAD_AXIS_LEFTX,
+		GAMEPAD_AXIS_LEFTY,
+		GAMEPAD_AXIS_RIGHTX,
+		GAMEPAD_AXIS_RIGHTY,
+		GAMEPAD_AXIS_TRIGGERLEFT,
+		GAMEPAD_AXIS_TRIGGERRIGHT,
+		GAMEPAD_AXIS_MAX_ENUM
+	};
+
+	// Valid Gamepad buttons.
+	enum GamepadButton
+	{
+		GAMEPAD_BUTTON_INVALID,
+		GAMEPAD_BUTTON_A,
+		GAMEPAD_BUTTON_B,
+		GAMEPAD_BUTTON_X,
+		GAMEPAD_BUTTON_Y,
+		GAMEPAD_BUTTON_BACK,
+		GAMEPAD_BUTTON_GUIDE,
+		GAMEPAD_BUTTON_START,
+		GAMEPAD_BUTTON_LEFTSTICK,
+		GAMEPAD_BUTTON_RIGHTSTICK,
+		GAMEPAD_BUTTON_LEFTSHOULDER,
+		GAMEPAD_BUTTON_RIGHTSHOULDER,
+		GAMEPAD_BUTTON_DPAD_UP,
+		GAMEPAD_BUTTON_DPAD_DOWN,
+		GAMEPAD_BUTTON_DPAD_LEFT,
+		GAMEPAD_BUTTON_DPAD_RIGHT,
+		GAMEPAD_BUTTON_MAX_ENUM
+	};
+
+	// Different types of inputs for a joystick.
+	enum InputType
+	{
+		INPUT_TYPE_AXIS,
+		INPUT_TYPE_BUTTON,
+		INPUT_TYPE_HAT,
+		INPUT_TYPE_MAX_ENUM
+	};
+
+	// Represents a gamepad input value, e.g. the "x" button or the left trigger.
+	struct GamepadInput
+	{
+		InputType type;
+		union
+		{
+			GamepadAxis axis;
+			GamepadButton button;
+		};
+	};
+
+	// Represents a joystick input value, e.g. button 6 or axis 1.
+	struct JoystickInput
+	{
+		InputType type;
+		union
+		{
+			int axis;
+			int button;
+			struct
+			{
+				int index;
+				Hat value;
+			} hat;
+		};
+	};
+
+	virtual ~Joystick() {}
+
+	virtual bool open(int deviceindex) = 0;
+	virtual void close() = 0;
+
+	virtual bool isConnected() const = 0;
+
+	virtual const char *getName() const = 0;
+
+	virtual int getAxisCount() const = 0;
+	virtual int getButtonCount() const = 0;
+	virtual int getHatCount() const = 0;
+
+	virtual float getAxis(int axisindex) const = 0;
+	virtual std::vector<float> getAxes() const = 0;
+	virtual Hat getHat(int hatindex) const = 0;
+
+	virtual bool isDown(const std::vector<int> &buttonlist) const = 0;
+
+	virtual bool openGamepad(int deviceindex) = 0;
+	virtual bool isGamepad() const = 0;
+
+	virtual float getGamepadAxis(GamepadAxis axis) const = 0;
+	virtual bool isGamepadDown(const std::vector<GamepadButton> &blist) const = 0;
+
+	virtual JoystickInput getGamepadMapping(const GamepadInput &input) const = 0;
+	virtual std::string getGamepadMappingString() const = 0;
+
+	virtual void *getHandle() const = 0;
+
+	virtual std::string getGUID() const = 0;
+	virtual int getInstanceID() const = 0;
+	virtual int getID() const = 0;
+
+	virtual void getDeviceInfo(int &vendorID, int &productID, int &productVersion) const = 0;
+
+	virtual bool isVibrationSupported() = 0;
+	virtual bool setVibration(float left, float right, float duration = -1.0f) = 0;
+	virtual bool setVibration() = 0;
+	virtual void getVibration(float &left, float &right) = 0;
+
+	static bool getConstant(const char *in, Hat &out);
+	static bool getConstant(Hat in, const char *&out);
+
+	static bool getConstant(const char *in, GamepadAxis &out);
+	static bool getConstant(GamepadAxis in, const char *&out);
+
+	static bool getConstant(const char *in, GamepadButton &out);
+	static bool getConstant(GamepadButton in, const char *&out);
+
+	static bool getConstant(const char *in, InputType &out);
+	static bool getConstant(InputType in, const char *&out);
+
+	static float clampval(float x);
+
+private:
+
+	static StringMap<Hat, HAT_MAX_ENUM>::Entry hatEntries[];
+	static StringMap<Hat, HAT_MAX_ENUM> hats;
+
+	static StringMap<GamepadAxis, GAMEPAD_AXIS_MAX_ENUM>::Entry gpAxisEntries[];
+	static StringMap<GamepadAxis, GAMEPAD_AXIS_MAX_ENUM> gpAxes;
+
+	static StringMap<GamepadButton, GAMEPAD_BUTTON_MAX_ENUM>::Entry gpButtonEntries[];
+	static StringMap<GamepadButton, GAMEPAD_BUTTON_MAX_ENUM> gpButtons;
+
+	static StringMap<InputType, INPUT_TYPE_MAX_ENUM>::Entry inputTypeEntries[];
+	static StringMap<InputType, INPUT_TYPE_MAX_ENUM> inputTypes;
+
+}; // Joystick
+
+} // joystick
+} // love
+
+
+#endif // LOVE_JOYSTICK_JOYSTICK_H
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/joystick/JoystickModule.h
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#ifndef LOVE_JOYSTICK_JOYSTICK_MODULE_H
+#define LOVE_JOYSTICK_JOYSTICK_MODULE_H
+
+// LOVE
+#include "common/Module.h"
+#include "Joystick.h"
+
+struct lua_State;
+
+namespace love
+{
+namespace joystick
+{
+
+class JoystickModule : public Module
+{
+public:
+
+	virtual ~JoystickModule() {}
+
+	// Implements Module.
+	ModuleType getModuleType() const override { return M_JOYSTICK; }
+
+	/**
+	 * Adds a connected Joystick device and opens it for use.
+	 * Returns NULL if the Joystick could not be added.
+	 **/
+	virtual Joystick *addJoystick(int deviceindex) = 0;
+
+	/**
+	 * Removes a disconnected Joystick device.
+	 **/
+	virtual void removeJoystick(Joystick *joystick) = 0;
+
+	/**
+	 * Gets a connected Joystick from its unique Instance ID.
+	 * Returns NULL if the instance ID does not correspond to a connected stick.
+	 **/
+	virtual Joystick *getJoystickFromID(int instanceid) = 0;
+
+	/**
+	 * Gets a connected Joystick.
+	 * Returns NULL if joyindex is not in the range of [0, getJoystickCount()).
+	 **/
+	virtual Joystick *getJoystick(int joyindex) = 0;
+
+	/**
+	 * Gets the index of a connected joystick.
+	 * Returns -1 if the joystick is not connected.
+	 **/
+	virtual int getIndex(const Joystick *joystick) = 0;
+
+	/**
+	 * Gets the number of currently connected Joysticks.
+	 **/
+	virtual int getJoystickCount() const = 0;
+
+	/**
+	 * Sets the virtual Gamepad mapping for a joystick input value for all
+	 * joystick devices with the specified joystick product GUID.
+	 * If any joysticks with the specified GUID are connected, they will be
+	 * added as Gamepads if they aren't already, otherwise their Gamepad mapping
+	 * will be updated.
+	 **/
+	virtual bool setGamepadMapping(const std::string &pguid, Joystick::GamepadInput gpinput, Joystick::JoystickInput joyinput) = 0;
+
+	/**
+	 * Loads a newline-separated list of virtual Gamepad mapping strings for
+	 * multiple joysticks at a time. The mapping strings must have been
+	 * generated with saveGamepadMappings, via Steam, or some other tool which
+	 * generates SDL GameController mappings.
+	 **/
+	virtual void loadGamepadMappings(const std::string &mappings) = 0;
+
+	/**
+	 * Gets a newline-separated list of virtual Gamepad mapping strings for
+	 * all used or modified Joysticks which are identified as Gamepads.
+	 **/
+	virtual std::string saveGamepadMappings() = 0;
+
+	/**
+	 * Gets the gamepad mapping string for the given GUID.
+	 **/
+	virtual std::string getGamepadMappingString(const std::string &guid) const = 0;
+
+}; // JoystickModule
+
+JoystickModule *newDoraJoystickModule(lua_State *L);
+
+} // joystick
+} // love
+
+#endif // LOVE_JOYSTICK_JOYSTICK_MODULE_H
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/joystick/wrap_Joystick.h
@@ -0,0 +1,40 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#ifndef LOVE_JOYSTICK_WRAP_JOYSTICK_H
+#define LOVE_JOYSTICK_WRAP_JOYSTICK_H
+
+// LOVE
+#include "common/config.h"
+#include "Joystick.h"
+#include "common/runtime.h"
+
+namespace love
+{
+namespace joystick
+{
+
+Joystick *luax_checkjoystick(lua_State *L, int idx);
+extern "C" int luaopen_joystick(lua_State *L);
+
+} // joystick
+} // love
+
+#endif // LOVE_JOYSTICK_SDL_WRAP_JOYSTICK_H
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/joystick/wrap_JoystickModule.h
@@ -0,0 +1,40 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#ifndef LOVE_JOYSTICK_WRAP_JOYSTICK_MODULE_H
+#define LOVE_JOYSTICK_WRAP_JOYSTICK_MODULE_H
+
+// LOVE
+#include "common/config.h"
+#include "common/runtime.h"
+#include "JoystickModule.h"
+
+namespace love
+{
+namespace joystick
+{
+
+int w_getIndex(lua_State *L);
+extern "C" LOVE_EXPORT int luaopen_love_joystick(lua_State *L);
+
+} // joystick
+} // love
+
+#endif // LOVE_JOYSTICK_WRAP_JOYSTICK_MODULE_H
--- a/Source/3rdParty/Love/src/modules/joystick/wrap_JoystickModule.cpp
+++ b/Source/3rdParty/Love/src/modules/joystick/wrap_JoystickModule.cpp
@@ -24,14 +24,12 @@
 #include "filesystem/Filesystem.h"
 #include "filesystem/wrap_Filesystem.h"
 
-#include "sdl/JoystickModule.h"
-
 namespace love
 {
 namespace joystick
 {
 
-#define instance() (Module::getInstance<JoystickModule>(Module::M_JOYSTICK))
+#define instance() (luax_getmodule<JoystickModule>(L, Module::M_JOYSTICK))
 
 int w_getJoysticks(lua_State *L)
 {
@@ -120,7 +118,7 @@
 	bool isfile = false;
 	std::string mappings = luax_checkstring(L, 1);
 
-	auto fs = Module::getInstance<love::filesystem::Filesystem>(Module::M_FILESYSTEM);
+	auto fs = luax_getmodule<love::filesystem::Filesystem>(L, Module::M_FILESYSTEM);
 	if (fs)
 	{
 		love::filesystem::Filesystem::Info info = {};
@@ -194,7 +192,7 @@
 	JoystickModule *instance = instance();
 	if (instance == nullptr)
 	{
-		luax_catchexcept(L, [&](){ instance = new sdl::JoystickModule(); });
+		luax_catchexcept(L, [&](){ instance = newDoraJoystickModule(L); });
 	}
 	else
 		instance->retain();

