--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/physics/Body.cpp
@@ -0,0 +1,59 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#include "Body.h"
+
+namespace love
+{
+namespace physics
+{
+
+love::Type Body::type("Body", &Object::type);
+
+Body::~Body()
+{
+}
+
+bool Body::getConstant(const char *in, Type &out)
+{
+	return types.find(in, out);
+}
+
+bool Body::getConstant(Type in, const char *&out)
+{
+	return types.find(in, out);
+}
+
+std::vector<std::string> Body::getConstants(Type)
+{
+	return types.getNames();
+}
+
+StringMap<Body::Type, Body::BODY_MAX_ENUM>::Entry Body::typeEntries[] =
+{
+	{"static", Body::BODY_STATIC},
+	{"dynamic", Body::BODY_DYNAMIC},
+	{"kinematic", Body::BODY_KINEMATIC},
+};
+
+StringMap<Body::Type, Body::BODY_MAX_ENUM> Body::types(Body::typeEntries, sizeof(Body::typeEntries));
+
+} // physics
+} // love
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/physics/Body.h
@@ -0,0 +1,63 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#ifndef LOVE_PHYSICS_BODY_H
+#define LOVE_PHYSICS_BODY_H
+
+// LOVE
+#include "common/Object.h"
+#include "common/StringMap.h"
+
+namespace love
+{
+namespace physics
+{
+
+class Body : public Object
+{
+public:
+
+	static love::Type type;
+
+	enum Type
+	{
+		BODY_INVALID,
+		BODY_STATIC,
+		BODY_DYNAMIC,
+		BODY_KINEMATIC,
+		BODY_MAX_ENUM
+	};
+
+	virtual ~Body();
+
+	static bool getConstant(const char *in, Type &out);
+	static bool getConstant(Type in, const char  *&out);
+	static std::vector<std::string> getConstants(Type);
+
+private:
+
+	static StringMap<Type, BODY_MAX_ENUM>::Entry typeEntries[];
+	static StringMap<Type, BODY_MAX_ENUM> types;
+};
+
+} // physics
+} // love
+
+#endif // LOVE_PHYSICS_BODY_H
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/physics/Shape.cpp
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#include "Shape.h"
+
+namespace love
+{
+namespace physics
+{
+
+love::Type Shape::type("Shape", &Object::type);
+
+Shape::~Shape()
+{
+}
+
+bool Shape::getConstant(const char *in, Type &out)
+{
+	return types.find(in, out);
+}
+
+bool Shape::getConstant(Type in, const char  *&out)
+{
+	return types.find(in, out);
+}
+
+StringMap<Shape::Type, Shape::SHAPE_MAX_ENUM>::Entry Shape::typeEntries[] =
+{
+	{"circle", Shape::SHAPE_CIRCLE},
+	{"polygon", Shape::SHAPE_POLYGON},
+	{"edge", Shape::SHAPE_EDGE},
+	{"chain", Shape::SHAPE_CHAIN},
+};
+
+StringMap<Shape::Type, Shape::SHAPE_MAX_ENUM> Shape::types(Shape::typeEntries, sizeof(Shape::typeEntries));
+
+} // physics
+} // love
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/physics/Shape.h
@@ -0,0 +1,63 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#ifndef LOVE_PHYSICS_SHAPE_H
+#define LOVE_PHYSICS_SHAPE_H
+
+// LOVE
+#include "common/Object.h"
+#include "common/StringMap.h"
+
+namespace love
+{
+namespace physics
+{
+
+class Shape : public Object
+{
+public:
+
+	static love::Type type;
+
+	enum Type
+	{
+		SHAPE_INVALID,
+		SHAPE_CIRCLE,
+		SHAPE_POLYGON,
+		SHAPE_EDGE,
+		SHAPE_CHAIN,
+		SHAPE_MAX_ENUM
+	};
+
+	virtual ~Shape();
+
+	static bool getConstant(const char *in, Type &out);
+	static bool getConstant(Type in, const char  *&out);
+
+private:
+
+	static StringMap<Type, SHAPE_MAX_ENUM>::Entry typeEntries[];
+	static StringMap<Type, SHAPE_MAX_ENUM> types;
+};
+
+} // physics
+} // love
+
+#endif // LOVE_PHYSICS_SHAPE_H
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/physics/Joint.cpp
@@ -0,0 +1,62 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#include "Joint.h"
+
+namespace love
+{
+namespace physics
+{
+
+love::Type Joint::type("Joint", &Object::type);
+
+Joint::~Joint()
+{
+}
+
+bool Joint::getConstant(const char *in, Type &out)
+{
+	return types.find(in, out);
+}
+
+bool Joint::getConstant(Type in, const char  *&out)
+{
+	return types.find(in, out);
+}
+
+StringMap<Joint::Type, Joint::JOINT_MAX_ENUM>::Entry Joint::typeEntries[] =
+{
+	{"distance", Joint::JOINT_DISTANCE},
+	{"revolute", Joint::JOINT_REVOLUTE},
+	{"prismatic", Joint::JOINT_PRISMATIC},
+	{"mouse", Joint::JOINT_MOUSE},
+	{"pulley", Joint::JOINT_PULLEY},
+	{"gear", Joint::JOINT_GEAR},
+	{"friction", Joint::JOINT_FRICTION},
+	{"weld", Joint::JOINT_WELD},
+	{"wheel", Joint::JOINT_WHEEL},
+	{"rope", Joint::JOINT_ROPE},
+	{"motor", Joint::JOINT_MOTOR},
+};
+
+StringMap<Joint::Type, Joint::JOINT_MAX_ENUM> Joint::types(Joint::typeEntries, sizeof(Joint::typeEntries));
+
+} // physics
+} // love
--- /dev/null
+++ b/Source/3rdParty/Love/src/modules/physics/Joint.h
@@ -0,0 +1,71 @@
+/**
+ * Copyright (c) 2006-2023 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#ifndef LOVE_PHYSICS_JOINT_H
+#define LOVE_PHYSICS_JOINT_H
+
+// LOVE
+#include "common/Object.h"
+#include "common/StringMap.h"
+
+namespace love
+{
+namespace physics
+{
+
+class Joint : public Object
+{
+public:
+
+	static love::Type type;
+
+	enum Type
+	{
+		JOINT_INVALID,
+		JOINT_DISTANCE,
+		JOINT_REVOLUTE,
+		JOINT_PRISMATIC,
+		JOINT_MOUSE,
+		JOINT_PULLEY,
+		JOINT_GEAR,
+		JOINT_FRICTION,
+		JOINT_WELD,
+		JOINT_WHEEL,
+		JOINT_ROPE,
+		JOINT_MOTOR,
+		JOINT_MAX_ENUM
+	};
+
+
+	virtual ~Joint();
+
+	static bool getConstant(const char *in, Type &out);
+	static bool getConstant(Type in, const char  *&out);
+
+private:
+
+	static StringMap<Type, JOINT_MAX_ENUM>::Entry typeEntries[];
+	static StringMap<Type, JOINT_MAX_ENUM> types;
+};
+
+} // physics
+} // love
+
+#endif // LOVE_PHYSICS_JOINT_H

